data-processing/entities/terms/extractor.py

```diff
@@ -10,110 +9,169 @@
+
+					# Object[0] is the term
+					if object[0] == word:
+
+						# Adding term name, text_segment from which term was found, def text and def html
+						matches_tuples.append([object[0], text_segment_string, object[1], object[2]])
+
+			for word2 in text_segment_string_not_split_on_dashes:
+				for object in terms_and_def_text_html:
+					if object[0] == word2:
+						matches_tuples.append([object[0], text_segment_string, object[1], object[2]])
+			for object in terms_and_def_text_html:
+				term_list = re.split("\s+", object[0])
+				if len(term_list) > 1:
+					if object[0] in text_segment_string:
+						matches_tuples.append([object[0], text_segment_string, object[1], object[2]])
+
+			for i in matches_tuples:
+				list_to_add = i[:-1]
+				if isinstance(i[-1], list):
+					last_elem = tuple(i[-1])
+				else:
+					last_elem = i[-1]
+				list_to_add.append(last_elem)
+				tuple_to_add = tuple(list_to_add)
+				matches.add(tuple_to_add)
+
+			###Getting the start and end indices of the matched terms
+			for match in matches:
+				start_index_of_string = text_segment_string.find(match[0])
+				end_index_of_string = start_index_of_string + len(match[0]) - 1
+				matched_terms.append(match[0])
+				matched_terms_context_str.append(match[1])
+				matched_definition_texts.append(match[2])
+				matched_definition_htmls.append(match[3])
+				matched_terms_start.append(text_segment.tex_start + start_index_of_string)
+				matched_terms_end.append(text_segment.tex_start + end_index_of_string)
+
+		def duplicates(lst, item):
+			return [i for i, x in enumerate(lst) if x == item]
+
+		# To Do:
+		# 1) Determine the objects that share the same start index.
+		# 2) Keep the object with the largest end index.
+		# 3) Remove all other objects from all lists.
+
+		def delete_duplicate_start(term, start, end, context_str):
+			for start_index in start:
+
+				# 1)
+				duplicate_indices = duplicates(start, start_index)
+
+				# 2) Keep the object with the largest end index.
+				if len(duplicate_indices) > 1:
+					print("OH NO")
+					ending_indices = []
+					for duplicate_index in duplicate_indices:
+						ending_indices.append(matched_terms_end[duplicate_index])
+
+					# get the index of the largest ending index
+					index_of_largest_ending_index = ending_indices.index(max(ending_indices))
+					duplicate_index_to_keep = index_of_largest_ending_index
+
+					# remove the index that we are going to keep from the list of duplicate indices.
+					duplicate_indices.pop(duplicate_index_to_keep)
+
+					# remove the objects that are located at the remaining indices from all lists.
+					for index in duplicate_indices:
+						matched_terms.pop(index)
+						matched_terms_start.pop(index)
+						matched_terms_end.pop(index)
+						matched_terms_context_str.pop(index)
+						matched_definition_texts.pop(index)
+						matched_definition_htmls.pop(index)
+
+		delete_duplicate_start(matched_terms, matched_terms_start, matched_terms_end, matched_terms_context_str)
```

Comment:
why would there be duplicate terms detected? This suggests to me that it might be worth refactoring the approach above to not have duplicates, to yield a simpler and more maintainable solution with fewer steps

Reply:
There technically aren't any "duplicate" terms; it's an issue with the glossary. The glossary has terms that are really similar: i.e., "convolutional neural networks", "convolutional neural network". The only difference is that one is plural and the other is singular (this is due to the glossary being compiled from different websites). I decided that it would be redundant to capture both of these (albeit they would have different definitions because they are listed in the glossary as different terms).

Reply:
How do you want me to handle this situation? I can capture both the singular and plural form, or just capture one of them. Currently, I am capturing the singular form

Reply:
Good catch! Can we just capture the singular form for each? Would there be any downsides to doing that?

Reply:
The code as is does that already; I'll leave it as is for now then. We might have to come back to this later, I thought of some different approaches that might be cleaner for the future (for example, checking if a word is in the glossary by checking once if the word (in it original form from the tex) is in the glossary, and then checking one more time if the word (after adding an "s" character to the end, making it the plural form) is in the glossary.

Reply:
I thought of some different approaches that might be cleaner (for example, checking if a word is in the glossary by checking once if the word (in it original form from the tex) is in the glossary, and then checking one more time if the word (after adding an "s" character to the end, making it the plural form) is in the glossary. This is something I am gonna come back to